### ISSUE TITLE
no-unused-prop-types with callback prop called in another callback

### ISSUE THREAD

**Issue by wbinnssmith** [yannickcr/eslint-plugin-react#1477]
The rule is flagging these as a false positive, though it is used.

Here's a minimal example:

```js
import React from 'react';

type Props = {
  test: string,
  callback: () => void,
};

export default function Foo(props: Props) {
  return (
    <div>
      {[1, 2, 3].map(e => (
        <div key={e} onClick={() => props.callback()}>
          {props.test}
        </div>
      ))}
    </div>
  );
}
```

**Comment by petersendidit** [comment]
I had a look in to this and was able to get some close but wasn't able to fix this completely.  I was able to add the correct handling for `CallExpression` but the real problem boils down to `component.getParentStatelessComponent` not being able to figure out what the actual component is. 

The current logic of `getParentStatelessComponent` decides that the arrow function provided as the callback to `map` is the component. This is hard to fix since with really it is a stateless functional component (it is a function that returns JSX).

`getParentStatelessComponent` correctly identifies `Foo` as component when `Components.detect` is called. A possible way to "fix" the detection of walking back up the scope, like in this case, would be to keep walking up arrow functions and see if we hit a component that we already know about.
